[DISCO/NUCLEO_L053xx] adding to RTOS, corrections for all tests, [DISCO_L053] exporter to µVision

ohagendorf opened the pull request:
Adding both targets to be able to run RTOS:
- changing default STM32Cube startup files to call _start instead of __libc_init_array and main
- extending RTX config and adding targets to RTOS_x tests in tests.py
- reducing stack sizes forthis target because of 8k RAM

Correcting the stdio uart config.

Solving the RTC start problem at DISCO_L053 target. Perhaps it is a hardware problem at the mcu when no 32kHz crystal is connected:
The mcu STM32L053C8 seems to have a problem in the RCC - LSE hardware block. The Disco_L053 don't have a 32kHz crystal connected to LSE port pins in contrast to NUCLEO_L053.
During initialization the HAL tests if it can start the LSE oscillator. The Flag LSERDY in RCC_CSR is set to 1 by RCC clock control when the oscillator runs stable. Without a crystal the flag shouldn't be set and the HAL would try to start the internal LSI oscillator.
But the flag is always set to 1 also without a crystal. 
That's why the RTC doesn't start. The suggested solution is starting the RTC directly with the internal LSI oscillator.

Every test (DTCT_1, EXAMPLE_1, MBED_xx, RTOS_x) is OK.

0xc0170 commented:
Can you rebase to the latest master please? There are few conflicts to resolve (I was not certain about resolving one of those so I stopped merging).

ohagendorf commented:
Hi,

I rebased it to the latest master.

0xc0170 commented on libraries/rtos/rtx/TARGET_CORTEX_M/cmsis_os.h:
```diff
@@ -121,7 +121,11 @@ used throughout the whole project.
 #endif
 
-#define DEFAULT_STACK_SIZE         (WORDS_STACK_SIZE*4)
+#if defined(TARGET_STM32L053R8) || defined(TARGET_STM32L053C8)
```
Why this is needed ? I found DEFAULT_STACK_SIZE used only in the Thread ctor, which has it as default parameter. So those 2 targets should define that parameter when created a new thread.

ohagendorf commented:
Without reducing the DEFAULT_STACK_SIZE not all test are running through. Of course it would also be possible to change the RTOS tests.

0xc0170 commented:
The application layer should change, what do you think

ohagendorf commented:
Yes, I think changing the tests is much better. And other tests already have target specific defines.
I'll add a commit.

ohagendorf commented:
I changed back cmsis_os.h to its original state and added the neccessary defines to the RTOS_x tests. All tests with an L053 and as a crosscheck with an F407 are fine.

0xc0170 commented:
I'll merge this as it is now. However, we should look at how to fix this properly - if different stack size is needed based on a target.

Can you create a new issue with a description - which tests are failing and why, which target and all information you have available. Thank you

ffxx68 commented:
Hi 0xc0170 , ohagendorf ,
I'm still getting "Error: Symbol SysTick_Handler multiply defined (by system_stm32l0xx.o and HAL_CM0.c.NUCLEO_L053R8.o)", while trying to compile the rtos_basic example.
I'm I missing some step? Is the merging complete?

ohagendorf commented:
@0xc0170 , @ffxx68 

I'm using the current master branch. My local git repo is at c:\github\mbed. When I've a commandline with the current folder workspace_tools and executing the following commands everything is OK:

```
C:\GitHub\mbed\workspace_tools>make.py -m NUCLEO_L053R8 -t GCC_ARM -p 77
Building project BASIC (NUCLEO_L053R8, GCC_ARM)
Compile: main.cpp
Link: basic
Elf2Bin: basic
Image: C:\GitHub\mbed\build\test\NUCLEO_L053R8\GCC_ARM\RTOS_1\basic.bin
```

and this generates complete projects for gcc_arm and coide:
```
C:\GitHub\mbed\workspace_tools>project.py -m NUCLEO_L053R8 -i coide -p 77
...

Successful exports:
  * NUCLEO_L053R8::coide        C:\GitHub\mbed\build\export\RTOS_1_coide_NUCLEO_L053R8.zip
```

```
C:\GitHub\mbed\workspace_tools>project.py -m NUCLEO_L053R8 -i gcc_arm -p 77
```

The first one can be build with:
```
C:\GitHub\mbed\build\export\RTOS_1_gcc_arm_NUCLEO_L053R8\RTOS_1>make
...
arm-none-eabi-size RTOS_1.elf
   text    data     bss     dec     hex filename
  16976     296    3664   20936    51c8 RTOS_1.elf
```

An the last one is successfully compiled by CoIDE:
```
...
       [cc] Starting link
       [cc] arm-none-eabi-gcc -mcpu=cortex-m0 -mthumb -g -Wl,-Map=RTOS_1.map -Os -Wl,--gc- ...
Program Size:
      text	   data	    bss	    dec	    hex	filename
     17908	    296	   3664	  21868	   556c	RTOS_1.elf

BUILD SUCCESSFUL
Total time: 19 seconds
```

ffxx68 commented:
Ok, so maybe it's because I'm using the online mbed IDE ... 
How is this propagated over there?
I'm sorry if my questions appear so basic (dumb?), but I'm new to this environment... as you might have guessed already.

ohagendorf commented:
OK, then it is clear why you get this error message.
The online compiler don't use the current offline tool version. From time to time the online system gets an update. So you have to wait a little bit. Or use the offline tools which works very well.

0xc0170 commented:
@ffxx68 always include in your report what version did you compile (for example - git commit SHA or mbed online revison)

mbed-src has nto been updated since December due an error.

ffxx68 commented:
I find the online IDE very handy... mbed-rtos over there is synced up to "DISCO_F401VC", which is  https://github.com/mbedmicro/mbed/commit/b6a514c8b2b7d3a778e1e644ca843be3c3f9facc
